Conversions of timeseries mass units, resampling and temporal homogenization

liviajakob opened the pull request:
Added some functionality to convert timeseries objects:

+ to different units (Gt and mwe), also including things such as change in area for example which is defined in the constants for each region.
+ to return an annual timeseries from higher and lower resolution timeseries
+ to return a longterm trend timeseries from higher resolution timeseries
+ a seasonally homogenized timeseries using a higher resolution calibration dataset

Now a Timeseries object can newly return a copy of itself with the converted changes. This is the design envisaged for any changes during the algorithm to each of the datasets. The idea behind that is that during the algorithm the actual data within a Timeseries object will not be edited, but we can return an edited new object. In this way we can compare/use the initial data at any point within the algorithm. Does this make sense?

Note that uncertainties are note implemented yet for this conversion, as the equation for it is not defined yet.

TristanGoss commented on glambie/data/data_catalogue.py:
```diff
@@ -53,19 +53,39 @@ def from_dict(meta_data_dict: dict) -> DataCatalogue:
             data catalogue containing the metadata of datasets, the actual timeseries data will lazy loaded
         """
-        basepath = os.path.join(*meta_data_dict['basepath'])
+        base_path = os.path.join(*meta_data_dict['basepath'])
```
This is a good change, can we change it in meta_data_dict as well?

liviajakob replied:
Yes, done

TristanGoss commented on glambie/data/timeseries.py:
```diff
@@ -201,3 +212,381 @@ def metadata_as_dataframe(self) -> pd.DataFrame:
+    def timeseries_is_annual_grid(self, year_type: str = "calendar"):
+        """
+        Returns True if all values in the self.data.start_dates and self.data.end_dates are on
+        the annual grid (e.g. 2010.0 wouls be on annual calendar year grid, 2010.1 would not)
```
typo - "wouls" should be "would"

liviajakob replied:
done

TristanGoss commented on glambie/data/timeseries.py:
```diff
@@ -201,3 +212,381 @@ def metadata_as_dataframe(self) -> pd.DataFrame:
+        """
+        Returns True if all values in the self.data.start_dates and self.data.end_dates are on
+        the monthly grid defined by glambie.util.timeseries_helpers.timeseries_as_months.
+        Also returns Trues if the resolution is not monthly, but all dates are using the monthly grid.
```
typo - Trues

liviajakob replied:
done

liviajakob commented on glambie/const/data_groups.py:
```diff
@@ -11,7 +11,7 @@ class GlambieDataGroup():
     'altimetry': GlambieDataGroup(name='altimetry', long_name='Altimetry'),
     'demdiff': GlambieDataGroup(name='demdiff', long_name='DEM differencing'),
     'gravimetry': GlambieDataGroup(name='gravimetry', long_name='Gravimetry'),
-    # Here, 'Glaciological' refers to measurements of the glacier's surface from seasonal or annual in-situ campaigns.
```
There will be a lot of documentation for these categories within the ATBD. Basically they are the main submission categories. So I see no need to explain the science behind it here, especially if it's only the term glaciological here. Then we'd have to explain all the others too. But I think the categories are clear for anyone who has been involved in the project.

TristanGoss replied:
If we want to create a situation where the code cannot be understood without the ATBD (which I would recommend against), then we should find a way to link the ATBD directly in the code, and we should consider referencing specific sections of the document, if it's longer than a few pages. The test for the codebase is, "can a fresh graduate understand this?" If the ATBD is linked and immediately accessible, perhaps. Without it, I think significant unecessary confusion would arise.

liviajakob replied:
I still think we can expect some understanding of the project itself here!

TristanGoss commented on glambie/data/timeseries.py:
```diff
@@ -201,3 +212,381 @@ def metadata_as_dataframe(self) -> pd.DataFrame:
+                object_copy.data.changes = np.array(changes)
+        return object_copy  # return copy of itself
+
+    def convert_timeseries_to_annual_trends(self, year_type="calendar") -> Timeseries:
```
It might be better to create an enum for the year type.

liviajakob replied:
How do you mean exactly? Like this: https://docs.python.org/3/library/enum.html within the constants?

liviajakob replied:
Thanks, Created a YearType(Enum) in constants.py

TristanGoss commented on glambie/data/timeseries.py:
```diff
@@ -201,3 +212,381 @@ def metadata_as_dataframe(self) -> pd.DataFrame:
+        seasonal_calibration_dataset : Timeseries
+            High resolution dataset to be used for calibration
+        year_type : str, optional
+            annual grid to which the tiemseries will be homogenized to, options are 'calendar', 'glaciological'
```
typeo: tiemseries

liviajakob replied:
done

TristanGoss commented on glambie/data/timeseries.py:
```diff
@@ -201,3 +212,381 @@ def metadata_as_dataframe(self) -> pd.DataFrame:
+            # 1) calibrate calibration series with trends from timeseries
+            calibrated_s, dist_mat = calibrate_timeseries_with_trends(self.data.as_dataframe(),
+                                                                      seasonal_calibration_dataset.data.as_dataframe())
+            # 2) caliculate mean calibration timeseries from all the different curves
```
typo - caliculate

liviajakob replied:
done

TristanGoss commented on glambie/util/mass_height_conversions.py:
```diff
@@ -215,6 +234,25 @@ def _gigatonnes_to_meters_water_equivalent(variable: float, area: float, density
     return (1e6 * variable) / (area * density_of_water)  # 1e6 to convert area from km2 to m2
 
 
+def _meters_water_equivalent_to_gigatonnes(variable: float, area: float, density_of_water: float = 997) -> float:
```
I don't think this needs to be a separate function, especially if it's private and called on only one line. Similarly with `_meters_water_equivalent_to_meters`. And I re-iterate my earlier comment about embedding units into variable names. Well done on typehinting `Iterable` though.

liviajakob replied:
I see your point, but that would mean changing it everywhere within this file (7 times), as they all have separate functions... I followed Sophies structure for mass unit conversions, not sure who reviewed that last time and proposed to do it that way

TristanGoss replied:
I would recommend changing it everywhere in the file. Doing so will halve the length of the file and make things easier to read.

As a general point, if you find yourself writing almost all of a docstring more than once, you probably don't have a good code structure.

TristanGoss commented on tests/data/test_timeseries.py:
```diff
@@ -122,3 +123,196 @@ def test_as_cumulative_timeseries_raises_warning(example_timeseries_ingested):
+    assert converted_timeseries.unit == "mwe"
+    assert example_timeseries_ingested.unit == "m"
+    assert not np.array_equal(converted_timeseries.data.changes, example_timeseries_ingested.data.changes)
+    expected_converted_changes = example_timeseries_ingested.data.changes / 997 * 850
```
It would be good to avoid magic numbers like these by creating a module containing well-named constants (for things like the density of water).

I've frequently had times where people have complained about the speed of light (because you get different numbers with different atmospheric models) and I've saved lots of time by being able to reach in and change it in just one place. In addition, Find-and-replace with numbers is significantly less reliable than doing it with long names.

TristanGoss replied:
For example here, 997 is the density of water at a particular temperature. Is that a sensible temperature for glaciers, ice sheets or ocean? Even if we say "yes", you can bet there's an academic somewhere that will say "well actually..."

liviajakob replied:
they are the default values in the function. could make it more obvious and pass them in.

liviajakob replied:
I would rather not make them constants as they need to stay flexible

liviajakob replied:
happier now? :)

TristanGoss replied:
So, the key goal is that the number "997", used to represent the density of water, should only occur once in the codebase. It should then be imported wherever it is used as a default setting.

There are two ways of doing this: The "Julia style", which is better for constants you think you'll almost never change, is to create a python module that only contains constants, so something like:

`constants.py` reads:
`DENSITY_OF_WATER_GT_PER_M3 = 997`

Then in other files:

```
import constants

def func(density_of_water = constants.DENSITY_OF_WATER_GT_PER_M3):
    stuff
```

The other way of doing it, let's call it the "EOLIS style", is to set it in a config file which is then read and distributed through the code as a config object of some kind. You can of course do both by instantiating the config object within constants.py, but this is rarely necessary.

But however it's achieved, for me this will be complete once the key goal is achieved: numbers that are constant should only occur once in the codebase, and I should never have to guess what a number I'm looking at represents (i.e. no "magic numbers").

liviajakob replied:
Thanks, went for Julia-method, and if we have configs in the end they could overwrite the const values?

TristanGoss replied:
yes, configs supplied as files would be expected to replace the values in the code, if the individual did not simply want to edit constants.py

TristanGoss reviewed (changes requested):
So normally I'm concerned about lots of uses of deepcopy because it burns memory, but here you appear to want to manipulate a timeseries the same way many people manipulate dataframes, which does make sense. Plus, the overall quantity of data is fairly small no matter what is done to it, so there shouldn't be any memory issues.

The algorithms make sense as well, I can see the adantage in seasonal homogenization.

I've left a few comments - I'd like you to address the ones I haven't already resolved. From memory, the main points are: "magic numbers" that are the kind of constants people regularly challenge / ask questions about need to be pulled out into a module, try to embed units into variable names, and a few typos.

liviajakob commented on glambie/data/timeseries.py:
```diff
@@ -201,3 +212,381 @@ def metadata_as_dataframe(self) -> pd.DataFrame:
+        elif rgi_area_version == 7:
+            glacier_area = self.region.rgi7_area
+        else:
+            warnings.warn("RGI version {} is not a valid version. As a default of version 6 is used."
```
Fair, changed it to raise an exception

liviajakob commented:
Thanks for the review! 
I addressed most of it (marked as resolved for the minor type changes and left the rest open)
I also had a couple of questions (see my comments)

liviajakob commented:
Thanks Tristan for clarifying my questions!
I think I addressed all your points now, but best to have another quick look over it from your side if you're happy with the changes. Good inputs, as always.

TristanGoss reviewed (approved):
Lovely, I'm happy with all of these new changes. Thanks for addressing my comments.
